### PR TITLE
hllplus: finish sparse (normalize + proto)

### PR DESCRIPTION
**Pull request by mxmCherry**

### CODE REVIEW

File: hllplus/hllplus_test.go [hunk 1]
```diff
@@ -238,34 +238,66 @@ var _ = Describe("HLL", func() {
...
+
+			// both precisions are always stored:
+			Expect(msg.GetPrecisionOrNumBuckets()).To(BeNumerically("==", 12))
+			Expect(msg.GetSparsePrecisionOrNumBuckets()).To(BeNumerically("==", 17))
+
+			// expect NO normal representation:
+			Expect(msg.GetData()).To(BeEmpty())
+
+			// expect sparse representation:
+			Expect(msg.GetSparseSize()).To(BeNumerically("==", 796)) // TODO: eh, shouldn't it be exactly 800???
```
Comment (mxmCherry) [inline hllplus/hllplus_test.go:290]: @dim please, pay attention to this question ^^

File: hllplus/sparse.go [hunk 2]
```diff
@@ -124,6 +152,33 @@ func (s *sparseState) encode(hash uint64) uint32 {
 	return s.encodedFlag | normPos<<sparseRhoWBits | uint32(rho)
 }
 
+// TODO: should we squash decodeNormalXXX into single func with 2 retvals?
```
Comment (dim) [inline hllplus/sparse.go:155]: yes, we should, much like `encode` we should have a single `decode`, the reason why there are two functions for it in Java is that it cannot do multiple return values

File: hllplus/sparse.go [hunk 3]
```diff
@@ -108,6 +117,25 @@ func (s *sparseState) OverMax() bool {
 	return s.data.Len() > s.maxDataLen
 }
 
+func (s *sparseState) Iterate(cb func(pos int, rhoW uint8)) {
```
Comment (dim) [inline hllplus/sparse.go:120]: consistency: I have changed `pos` to be `uint32` everywhere

File: hllplus/sparse.go [hunk 4]
```diff
@@ -124,6 +152,33 @@ func (s *sparseState) encode(hash uint64) uint32 {
 	return s.encodedFlag | normPos<<sparseRhoWBits | uint32(rho)
 }
 
+// TODO: should we squash decodeNormalXXX into single func with 2 retvals?
+func (s *sparseState) decodeNormalIndex(sparseValue uint32) int {
```
Comment (dim) [inline hllplus/sparse.go:156]: return uint32 here too

File: hllplus/sparse.go [hunk 5]
```diff
@@ -124,6 +152,33 @@ func (s *sparseState) encode(hash uint64) uint32 {
...
+	// Values without a sparse rhoW' consist of just the sparse index, so the normal index is
+	// determined by stripping off the last sp-p bits.
+	if sparseValue&s.encodedFlag == 0 {
+		return int(sparseValue >> (s.sparsePrecision - s.normalPrecision))
```
Comment (dim) [inline hllplus/sparse.go:160]: no need to cast to int

File: hllplus/sparse.go [hunk 6]
```diff
@@ -124,6 +152,33 @@ func (s *sparseState) encode(hash uint64) uint32 {
...
+	// off the front and the rhoW' off the end.
+	return int((sparseValue ^ s.encodedFlag) >> sparseRhoWBits)
+}
+
+// TODO: should we squash decodeNormalXXX into single func with 2 retvals?
+func (s *sparseState) decodeNormalRhoW(sparseValue uint32) uint8 {
+	// If the rhoW' was not encoded, we can determine the normal rhoW from the last sp-p bits of
+	// the sparse index.
+	if (sparseValue & s.encodedFlag) == 0 {
+		// TODO: is it Ok just to cast `sparseValue` uint32 -> uint64?
```
Comment (dim) [inline hllplus/sparse.go:173]: yes, OK, pls remove TODO

File: hllplus/sparse.go [hunk 7]
```diff
@@ -254,3 +309,18 @@ func (s *deltaSlice) Iterate(fn func(uint32)) {
 		last = x
 	})
 }
+
+func (s *deltaSlice) CopyBytes() []byte {
```
Comment (dim) [inline hllplus/sparse.go:313]: we use `Clone` everywhere, but OK

File: hllplus/hllplus_test.go (hunk 1 above)
Comment (dim) [reply hllplus/hllplus_test.go:290]: so it depends really, each value we add to sparse is first encoded within sparse precision `(1<<17)`, i.e. `0 .. 131072` and added to `s.data.nums`. which means that you can have collisions. if you choose a larger sparse precision above, e.g. `20`, does this number change?

File: hllplus/sparse.go (hunk 7 above)
Comment (mxmCherry) [reply hllplus/sparse.go:313]: Well, Clone copies entire receiver; I wanted smth to get an immediate value of it's buffer.
Or you mean, let's do `deltaSlice.Clone().Bytes()`?

Comment (dim) [reply hllplus/sparse.go:313]: I get that, I think the overhead `deltaSlice.Clone().Bytes()` is tiny and it's a) easier to read and b) less code needed

File: hllplus/hllplus_test.go (hunk 1 above)
Comment (mxmCherry) [reply hllplus/hllplus_test.go:290]: @dim 
- sparsePrecision=17 -> GetSparseSize=798
- sparsePrecision=20 -> GetSparseSize=799

Will add a comment and leave as is.

File: hllplus/hllplus.go [hunk 8]
```diff
@@ -197,6 +199,8 @@ func (s *HLL) Downgrade(precision, sparsePrecision uint8) error {
 		return err
 	}
 
+	// TODO: downgrade sparse as well (and don't forget a switch between normal and sparse)
```
Comment (mxmCherry) [inline hllplus/hllplus.go:202]: Is it "downgrade-able"?

Comment (dim) [reply hllplus/hllplus.go:202]: yes, process is pretty much the same, something for later

**Comment by dim** [comment]
Sorry, one more unrelated thing that I forgot: `downgradeEach` should also work with `pos uint32` not `pos int`, otherwise :+1:

**Comment by dim** [comment]
Oh and pls fix CI

**Pull request merged**